docs/WritingTests.asciidoc

```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+test designer job_ (beware deadlocks).
+
+In short, writing multi-machine tests add few more layers of complexity:
+
+1. documenting the dependencies and order between individual tests
+2. synchronization between individual tests
+3. actual technical realization (i.e. https://github.com/os-autoinst/os-autoinst/blob/master/doc/networking.md[custom networking])
+
+Job dependencies
+~~~~~~~~~~~~~~~~
+
+There are 2 types of dependencies: +CHAINED+ and +PARALLEL+:
+
+* +CHAINED+ describes when one test case depends on another and are run sequentially, i.e. KDE test suite is run after and only after Installation test suite
+is successfully finished and canceled if fail.
+
+To define +CHAINED+ dependency add variable +START_AFTER_TEST+ with the name(s) of test suite(s) after which selected test suite is supposed to run.
+
+* +PARALLEL+ describes MM test, test suites are run at the same time and managed as a group.
+
+To define +PARALLEL+ dependency, use +PARALLEL_WITH+ variable with the name(s) of test suite(s) which acts as a parent suite to selected test suite.
+
+Worker requirements
+^^^^^^^^^^^^^^^^^^^
+
+Beware OpenQA worker requirements. +CHAINED+ dependency requires only one worker, since dependent jobs will run only after the first one finish.
+On the other hand +PARALLEL+ dependency requires at least 2 workers for simple scenarios.
+
+Keep in mind parent job _must be running until all
+children finish_, else scheduler will cancel child jobs once parent is done.
+
+
+Examples:
+^^^^^^^^^
+
+.+CHAINED+ - i.e. test basic functionality before going advanced - requires 1 worker
+----
+A <- B <- C
+
+Define test suite A,
+then define B with variable START_AFTER_TEST=A and then define C with START_AFTER_TEST=B
+
+-or-
+
+Define test suite A, B
+and then define C with START_AFTER_TEST=A,B
+In this case however the start order of A and B is not specified.
+But C will start only after A, B are successfully done.
+----
+.+PARALLEL+ with multiple parents - i.e. complex support requirements for one test - requires 4 workers
+----
+A <-\
+B <--- D
+C <-/
+
+Define test suites A,B,C
+and then define D with PARALLEL_WITH=A,B,C
+----
+.+PARALLEL+ with one parent - i.e. running independent tests against one server - requires at least 2 workers
+----
+    /- B
+A <--- C
+    \- D
+
+Define test suite A
+and then define B,C,D with PARALLEL_WITH=A
+----
+Test synchronization and locking API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+OpenQA provides locking server through lock API. To use lock API import +lockapi+ package (_use lockapi;_) in your test file.
+Lock API provides three functions: +mutex_create+, +mutex_lock+, +mutex_unlock+. Each of these functions take one parameter, name of the lock.
+
+By default, when mutex lock does not exist, lock is considered locked and +mutex_lock+ blocks. After lock is created by +mutex_create+, lock
+is automatically unlocked. +mutex_unlock+ unlocks the lock.
+
```

Comment:
There was an idea to use mutex_unlock as a form of blocking is_locked mechanism. So that it blocks when locked, but after unlocked it does not change the lock. Unfortunately as I go through the code this does not work, yet. Will fix.
The following example illustrates exactly that ... not. :(